Show bucket:create permission in permissions endpoint (ref #1510)

stloma opened the pull request:
Fixes #1510 

For the `kinto/core/utils.py` change, `''` is required for the `resource_name` value in the dictionary lookup in the permissions tree:

https://github.com/Kinto/kinto/blob/6d77dcb2b94f9198fa08db489776440f7df0c275/kinto/views/permissions.py#L123

Since the purpose of the `view_lookup` function is to map a `uri` to a `resource_name`, I thought it best in there.

I was also going under the assumption that the permissions response for the top bucket should not return an id, so there are changes to that effect as well.

Everything else should be pretty straightforward and please let me know if you see any issues or suggestions.

Thanks!

leplatrem commented on kinto/core/utils.py:
```diff
@@ -445,6 +445,11 @@ def view_lookup(request, uri):
+
+    # If we are accessing the top level listing of /buckets, there is no id
+    if (uri == '/buckets'):
+        return '', {id: False}
```
I'm pretty sure `view_lookup(request, '/buckets')` should already work as expected, ie. return `"bucket", {}`

If it does not, then we can fix it. You could add a unit test in test_utils to assert that BTW

stloma replied:
Yes `view_lookup` does return `"bucket", {}` without that code, the issue I'm seeing is that I need it to return `'', {}` for this:

https://github.com/Kinto/kinto/blob/6d77dcb2b94f9198fa08db489776440f7df0c275/kinto/views/permissions.py#L123

`perms_descending_tree` is populated with:

https://github.com/Kinto/kinto/blob/6d77dcb2b94f9198fa08db489776440f7df0c275/kinto/authorization.py#L26

So I need `''` for `resource_name` in the `obtained = perms_descending_tree[resource_name][perm]` code above.

That's what's causing some of the back and forth between assigning `bucket` and `''` in some of my other changes.

leplatrem replied:
Ok, I get it.

So we should keep the `view_lookup()` consistent with what it does. `/buckets` is the `bucket` resource, so it should return.

Now, what we want is that to read `perms_descending_tree[""]["bucket:create"]`. So we should manage this corner here, and not in `view_lookup()`.

For example:
```py
if resource == "bucket" and permission == "create":
    obtained = perms_descending_tree[""]["bucket:create"]
```

That way we don't have to change `kinto.authorization.PERMISSIONS_INHERITANCE_TREE`, and we dont' change the semantics of `view_lookup()` for that particular. Just imagine that `view_lookup()` can be used by any other component and would not expect this `"", {"id": False}` response

leplatrem commented on kinto/views/permissions.py:
```diff
@@ -114,8 +118,12 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+            if matchdict.get('id'):
+                matchdict[resource_name + '_id'] = matchdict.get('id')
+            else:
+                matchdict = {}
```
this would be simpler like this:
```py
if "id" in matchdict:
     matchdict[resource_name + '_id'] = matchdict['id']
```

stloma replied:
Yes absolutely, thanks!

leplatrem commented on kinto/views/permissions.py:
```diff
@@ -125,6 +133,11 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+                # For consistency with event payloads, change top level bucket
+                # resource_name from '' to bucket
+                if resource_name == '':
+                    resource_name = 'bucket'
```
this would not be necessary if `view_lookup()` would return the right value

stloma replied:
From my comment on `kinto/core/utils.py`, I need `''` to get the permissions in:

https://github.com/Kinto/kinto/blob/6d77dcb2b94f9198fa08db489776440f7df0c275/kinto/views/permissions.py#L123

Then I have to switch it back to `bucket` for the response payload the user sees. If I didn't make this change, the response would be:

```
"data": [
        {
            "permissions": [
                "bucket:create"
            ],
            "resource_name": "",
            "uri": "/buckets"
        }
    ]
```

Rather than:

```
"data": [
        {
            "permissions": [
                "bucket:create"
            ],
            "resource_name": "bucket",
            "uri": "/buckets"
        }
    ]
```

leplatrem commented on kinto/authorization.py:
```diff
@@ -25,7 +25,9 @@
-        'bucket:create': {},
+        'bucket:create': {
+            '': ['bucket:create', 'create']
+        },
```
Are you sure this is necessary? I would say that the code in `from_settings()` is enough.

leplatrem commented on tests/test_views_permissions.py:
```diff
@@ -19,6 +19,44 @@ def get_app_settings(cls, extras=None):
+        return settings
+
+
+class EntriesTestUnauthenticated(PermissionsUnauthenticatedViewTest):
```
You don't need that class, you can just put the test into `PermissionsUnauthenticatedViewTest`

stloma replied:
Thanks, this was my first time doing real testing, I'll clean up all of that.

leplatrem commented on tests/test_views_permissions.py:
```diff
@@ -19,6 +19,44 @@ def get_app_settings(cls, extras=None):
+                          headers=get_user_headers('alice'))
+
+    def test_permissions_list_entries_for_everyone(self):
+        print(self.headers)
```
leftover

leplatrem commented on tests/test_views_permissions.py:
```diff
@@ -19,6 +19,44 @@ def get_app_settings(cls, extras=None):
+        print(self.headers)
+        resp = self.app.get('/permissions', headers=self.everyone_headers)
+        permissions = resp.json['data']
+        self.assertEqual(len(permissions), 5)
```
There needs to be a test somewhere that asserts the content of the bucket creation entry

stloma replied:
Will do

leplatrem commented on tests/test_views_permissions.py:
```diff
@@ -168,6 +206,8 @@ def test_permissions_granted_via_groups_are_listed(self):
         resp = self.app.get('/permissions', headers=self.admin_headers)
         collections = [e for e in resp.json['data']
+                       # top level '/buckets' does not have an id
+                       if e['uri'] != '/buckets'
```
syntax error?

stloma replied:
This is due to getting a KeyError since the top level bucket doesn't have an id

Natim commented on kinto/authorization.py:
```diff
@@ -25,7 +25,9 @@
     '': {  # Granted via settings only.
-        'bucket:create': {},
+        'bucket:create': {
+            '': ['bucket:create', 'create']
```
It doesn't makes sense to inherit `bucket:create` from `bucket:create` because we already have it.

leplatrem replied:
No, I don't think this `PERMISSIONS_INHERITANCE_TREE` dict should be touched in this PR. 
This dict gives the permission inheritance, and there is permission that is above `bucket:create` since it can be given via settings.

However in the permissions endpoint code, we can manipulate the `perms_descending_tree` and adjust the `from_settings()` function.

Natim replied:
Oh this is because we don't inherit settings permission just yet ok...

Natim commented:
Thank you for tackling this :+1:

stloma commented:
Hey Natim, sure thing!

Sorry, I have a silly question, how do I push my changes here?

I've made changes on my local branch and read that to push them to a pull request, I would do a commit --amend, push them to my my branch, and they would automatically integrate into this pull request, but it looks like they're just showing as outdated here. What is the best method for this?

Natim commented:
> Sorry, I have a silly question, how do I push my changes here?


You can keep committing to your branch and it will automatically update the PR.

stloma commented:
Ok great, thanks. I'm just not seeing the updated code here.

In any case, I was able to cleanup the code and a solution to the `PERMISSIONS_INHERITANCE_TREE` if you all think that it works.

I also took out the `view_lookup` code, thanks a lot for the explanation on that leplatrem, it made a lot of sense.

If you all have any more suggestions, please let me know, otherwise I'll start the test cleanup.

Natim commented:
> I'm just not seeing the updated code here.

I can't see it there too: https://github.com/stloma/kinto/tree/show-bucket-create
So I believe your push wasn't successful.

stloma commented:
Hmm, I'm seeing it there. I took out the `bucket:create` perm from `''` here for instance:

https://github.com/stloma/kinto/blob/fa9652583cff52233c022cb864f12ef3da8a4466/kinto/authorization.py#L27-L29

Natim commented:
This commit, https://github.com/stloma/kinto/commit/fa9652583cff52233c022cb864f12ef3da8a4466 is the one that is linked to that PR and that you pushed force. (It updated the PR too btw) so we are good. What misled me is the fact that it still has the time of the initial commit. Try to avoid using `git commit --amend``

stloma commented:
Will do, I'll just do a regular commit next time with the same message, thanks for the help :)

leplatrem commented on kinto/views/permissions.py:
```diff
@@ -82,6 +82,13 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+        bucket_creation = from_settings.get('')
+        if bucket_creation:
+            perms_by_object_uri['/buckets'] = bucket_creation
+            perms_descending_tree[''] = {'bucket:create': {'': {'bucket:create'}}}
```
Why not do `perms_descending_tree['bucket'] = {'bucket:create': {'bucket': {'bucket:create'}}}` here and remove the if below ?

stloma replied:
The issue I saw with that is the way you wrote it, `perms_descending_tree['bucket'] = {'bucket:create': {'bucket': {'bucket:create'}}}`, would overwrite the existing `bucket` permissions taken in from `PERMISSIONS_INHERITANCE_TREE`

We could do something like `perms_descending_tree['bucket'][''] = {'bucket:create': {'': {'bucket:create'}}}`, but then `obtained = perms_descending_tree[resource_name][perm]` would get messed up and we'd be back to another if statement.

stloma replied:
No, you're right, I can do the following and remove the if statement:

`perms_descending_tree['bucket'].update({'bucket:create': {'bucket': {'bucket:create'}}})`

stloma replied:
Just let me know if that's acceptable and I'll start on the tests.

leplatrem replied:
yep!

stloma commented on kinto/views/permissions.py:
```diff
@@ -82,12 +82,14 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+                perms_by_object_uri['/buckets'] = {bucket_perm}
+                perms_descending_tree['bucket'].update({
+                    bucket_perm: {'bucket': {bucket_perm}}
+                })
```
I found that the accounts plugin also populates `from_settings` with `account:create`, so I added a for loop as well that will allow for multiple settings.

leplatrem replied:
Oh! Very good catch!

I think we'll have to do the same thing for accounts, I'm pretty sure the permission to create and write account does not appear in the permission endpoint. We should open an issue and fix it in another PR.

stloma commented:
Actually, hold off on this commit if you can :(

I've been looking into displaying if the user has account:create permissions for the /permissions endpoint that you mentioned and think I found a bug.

stloma commented:
Ok, fixed :)

I thought with both the accounts plugin and `kinto.bucket_create_principals = system.Everyone` it was working, but there was an issue where `account:create` was applying to the `/buckets` uri. I fixed that and adding more testing so it wouldn't happen again.

By doing it this way, the `/accounts` uri and `/accounts/{username}` is now showing on the permissions endpoint without creating an additional issue and without going outside of the scope of fixing the original issue that this pull request references.

It should be good to go!

leplatrem commented on CHANGELOG.rst:
```diff
@@ -15,6 +15,7 @@ This document describes changes between each past release.
 - Now fully rely on Pyup.io (or contributors) to update the versions in the `requirements.txt` file (fixes #1512)
 - Move from importing pip to running it in a subprocess (see https://github.com/pypa/pip/issues/5081).
 - Remove useless print when using the OpenID policy (ref #1509)
+- Add creation of buckets in permissions endpoint (ref #1510)
```
And accounts ;)

leplatrem commented on kinto/views/permissions.py:
```diff
@@ -82,6 +82,17 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+        # Add additional resources and permissions defined in settings/plugins
+        bucket_perms = from_settings.get('')
+        if bucket_perms:
+            for bucket_perm in bucket_perms:
```
This should probably be renamed `root_perms` now.
And I guess it can be simplified:
```py
for root_perm in from_settings.get('', []):
    resource_name, _ = ...
```

leplatrem commented on kinto/views/permissions.py:
```diff
@@ -82,6 +82,17 @@ def get_records(self, filters=None, sorting=None, pagination_rules=None,
+        if bucket_perms:
+            for bucket_perm in bucket_perms:
+                resource_name, _ = bucket_perm.split(':')
+                uri = '/{0}s'.format(resource_name)
```
This will work, but if we want to be completely safe here, it should be:
```py
uri = core_utils.strip_uri_prefix(self.request.route_path(f'{resource_name}-collection'))
```

leplatrem commented on tests/plugins/test_accounts.py:
```diff
@@ -252,7 +253,52 @@ def test_permissions_endpoint_is_compatible_with_accounts_plugin(self):
+        self.assertEqual(accounts['id'], 'alice')
+        self.assertEqual(account['permissions'], ['account:create'])
+        self.assertEqual(account['resource_name'], 'account')
+        self.assertEqual(account['uri'], '/accounts')
```
Why `account` singular if it is for `/accounts`?

stloma replied:
I was following the naming convention for `resource_name`, i.e., `bucket`, `collection`, `group`, and `record`, so thought `account` would be appropriate for the `uri` `/accounts`.

Thanks for the rest of the input too, I'll incorporate them. I'll just wait for your thoughts on `account` vs `accounts` for `resource_name` and then send the changes.

leplatrem replied:
ok I get it ! leave it like that sorry :)

leplatrem commented on tests/plugins/test_accounts.py:
```diff
@@ -252,7 +253,52 @@ def test_permissions_endpoint_is_compatible_with_accounts_plugin(self):
+        buckets = resp.json['data']
+        accounts = buckets[4]
+        account = buckets[5]
+        self.assertEqual(len(accounts), 5)
```
Here you test the number of keys, I think it would be just easier to compare the dict directly:
```py
self.assertEqual(account, {
    'permissions': ['read', 'write'],
    'resource_name', ...
    ...
})
```

leplatrem commented on tests/plugins/test_accounts.py:
```diff
@@ -252,7 +253,52 @@ def test_permissions_endpoint_is_compatible_with_accounts_plugin(self):
+        bucket_create = buckets[0]
+        account_create = buckets[1]
+        self.assertEqual(len(bucket_create), 3)
+        self.assertEqual(len(account_create), 3)
```
Same here, I think you can compare dicts

leplatrem commented on tests/test_views_permissions.py:
```diff
@@ -19,6 +19,59 @@ def get_app_settings(cls, extras=None):
+        self.assertEqual(len(toplevel_bucket), 3)
+        self.assertEqual(toplevel_bucket['permissions'], ['bucket:create'])
+        self.assertEqual(toplevel_bucket['resource_name'], 'bucket')
+        self.assertEqual(toplevel_bucket['uri'], '/buckets')
```
Compare dicts here as well?

stloma commented:
Hey leplatrem, sorry, sick of me yet? :)

My latest push is hitting a test failure that I didn't pick up on my local machine. Where we changed one of my tests to test by dictionary comparison:

```
self.assertEqual(accounts, {
    ...
   'permissions': ['write', 'read'],
  ...
})
```

It is hitting a sort of race condition (if that's the correct term) since the value of the `permissions` key is a set, thus unordered, so sometimes I get `'permissions': ['write', 'read']`, other times `'permissions': ['read','write']`. These values are populated by:

https://github.com/Kinto/kinto/blob/7ea5061c6dd3e8a452be0d901948ab55f14e18c3/kinto/views/permissions.py#L73

So all permissions displayed on the permissions endpoint, e.g., `buckets`, `collections`, etc, will come out in different orders.

For example:

```
 "bucket_id": "mybucket", 
            ....
            "permissions": [
                "read", 
                "collection:create", 
                "group:create", 
                "write", 
                "read:attributes"
            ], 
```

vs.

```
 "bucket_id": "mybucket", 
            ....
            "permissions": [
                "collection:create", 
                "read", 
                "read:attributes",
                "group:create", 
                "write"    
            ], 
```


I don't know if this is an issue or not, or whether I can just go ahead and change my test so it deals with it, I just wanted to bring it to your attention before I moved forward.

Also, please correct me if there's a better way to handle these issues that are coming up, I'm not trying to be a bother :)

glasserc commented:
Hi @stloma, I can't speak for @leplatrem but I took a few minutes to look through the other tests in the codebase and there seem to be a few in `test_object_permissions.py` that use `sorted()` to compare, probably for exactly this reason. I think that's the best approach. It's not exactly a race condition because it doesn't have to do with two processes happening at the same time, just that we use a `set()` so even within one process, order isn't guaranteed. JSON doesn't have any way to represent a `set()` except as a JSON Array, so it looks like there's an order when we really don't intend for there to be.

Don't worry about asking questions or leaving comments on a PR! We like to work collaboratively! Thanks for your continued hard work!

stloma commented:
Awesome, thanks for your help!
